diff --git a/libass/ass_bitmap.c b/libass/ass_bitmap.c
index d3186b4..53e4c6a 100644
--- a/libass/ass_bitmap.c
+++ b/libass/ass_bitmap.c
@@ -42,7 +42,7 @@
 #undef ALIGN
 #undef DECORATE
 
-#if (defined(__i386__) || defined(__x86_64__)) && CONFIG_ASM
+#if (defined(__i386__) || defined(__x86_64__) || defined(_M_IX86) || defined(_M_X64)) && CONFIG_ASM
 
 #define ALIGN           4
 #define DECORATE(func)  ass_##func##_sse2
diff --git a/libass/ass_compat.h b/libass/ass_compat.h
index ae16bb7..310a7e7 100644
--- a/libass/ass_compat.h
+++ b/libass/ass_compat.h
@@ -24,6 +24,10 @@
 #define _CRT_SECURE_NO_WARNINGS
 #define _USE_MATH_DEFINES
 #define inline __inline
+#include <crtversion.h>
+#if _VC_CRT_MAJOR_VERSION < 14
+#define snprintf _snprintf
+#endif
 #endif
 
 #ifndef HAVE_STRDUP
diff --git a/libass/ass_directwrite.c b/libass/ass_directwrite.c
index e666bd9..7819724 100644
--- a/libass/ass_directwrite.c
+++ b/libass/ass_directwrite.c
@@ -22,13 +22,16 @@
 
 #include <initguid.h>
 #include <ole2.h>
-#include <shobjidl.h>
 
 #include "dwrite_c.h"
 
 #include "ass_directwrite.h"
 #include "ass_utils.h"
 
+#if defined(WINAPI_FAMILY) && (WINAPI_FAMILY==WINAPI_FAMILY_PC_APP || WINAPI_FAMILY==WINAPI_FAMILY_PHONE_APP)
+#define WINRT
+#endif
+
 #define NAME_MAX_LENGTH 256
 #define FALLBACK_DEFAULT_FONT L"Arial"
 
@@ -48,7 +51,9 @@ typedef struct {
 } FontPrivate;
 
 typedef struct {
+#ifndef WINRT
     HMODULE directwrite_lib;
+#endif
     IDWriteFactory *factory;
     IDWriteGdiInterop *gdi_interop;
 } ProviderPrivate;
@@ -392,7 +397,9 @@ static void destroy_provider(void *priv)
     ProviderPrivate *provider_priv = (ProviderPrivate *)priv;
     provider_priv->gdi_interop->lpVtbl->Release(provider_priv->gdi_interop);
     provider_priv->factory->lpVtbl->Release(provider_priv->factory);
+#ifndef WINRT
     FreeLibrary(provider_priv->directwrite_lib);
+#endif
     free(provider_priv);
 }
 
@@ -760,6 +767,9 @@ ASS_FontProvider *ass_directwrite_add_provider(ASS_Library *lib,
     DWriteCreateFactoryFn DWriteCreateFactoryPtr = NULL;
     ProviderPrivate *priv = NULL;
 
+#ifdef WINRT
+    DWriteCreateFactoryPtr = DWriteCreateFactory;
+#else
     HMODULE directwrite_lib = LoadLibraryW(L"Dwrite.dll");
     if (!directwrite_lib)
         goto cleanup;
@@ -769,6 +779,7 @@ ASS_FontProvider *ass_directwrite_add_provider(ASS_Library *lib,
                                                       "DWriteCreateFactory");
     if (!DWriteCreateFactoryPtr)
         goto cleanup;
+#endif
 
     hr = DWriteCreateFactoryPtr(DWRITE_FACTORY_TYPE_SHARED,
                                 &IID_IDWriteFactory,
@@ -791,7 +802,9 @@ ASS_FontProvider *ass_directwrite_add_provider(ASS_Library *lib,
     if (!priv)
         goto cleanup;
 
+#ifndef WINRT
     priv->directwrite_lib = directwrite_lib;
+#endif
     priv->factory = dwFactory;
     priv->gdi_interop = dwGdiInterop;
 
@@ -808,8 +821,10 @@ cleanup:
         dwGdiInterop->lpVtbl->Release(dwGdiInterop);
     if (dwFactory)
         dwFactory->lpVtbl->Release(dwFactory);
+#ifndef WINRT
     if (directwrite_lib)
         FreeLibrary(directwrite_lib);
+#endif
 
     return NULL;
 }
diff --git a/libass/ass_render.c b/libass/ass_render.c
index 3b707a7..acd64b0 100644
--- a/libass/ass_render.c
+++ b/libass/ass_render.c
@@ -74,7 +74,7 @@ ASS_Renderer *ass_renderer_init(ASS_Library *library)
     priv->ftlibrary = ft;
     // images_root and related stuff is zero-filled in calloc
 
-#if (defined(__i386__) || defined(__x86_64__)) && CONFIG_ASM
+#if (defined(__i386__) || defined(__x86_64__) || defined(_M_IX86) || defined(_M_X64)) && CONFIG_ASM
     if (has_avx2())
         priv->engine = &ass_bitmap_engine_avx2;
     else if (has_sse2())
diff --git a/libass/ass_utils.c b/libass/ass_utils.c
index 65b6f74..a0a74cc 100644
--- a/libass/ass_utils.c
+++ b/libass/ass_utils.c
@@ -31,7 +31,7 @@
 #include "ass_utils.h"
 #include "ass_string.h"
 
-#if (defined(__i386__) || defined(__x86_64__)) && CONFIG_ASM
+#if (defined(__i386__) || defined(__x86_64__) || defined(_M_IX86) || defined(_M_X64)) && CONFIG_ASM
 
 #include "x86/cpuid.h"
 
diff --git a/libass/ass_utils.h b/libass/ass_utils.h
index 377d532..98f329d 100644
--- a/libass/ass_utils.h
+++ b/libass/ass_utils.h
@@ -48,7 +48,7 @@
 
 #define ASS_PI 3.14159265358979323846
 
-#if (defined(__i386__) || defined(__x86_64__)) && CONFIG_ASM
+#if (defined(__i386__) || defined(__x86_64__) || defined(_M_IX86) || defined(_M_X64)) && CONFIG_ASM
 int has_sse2(void);
 int has_avx(void);
 int has_avx2(void);
diff --git a/libass/dwrite_c.h b/libass/dwrite_c.h
index 5612e9f..e868ae9 100644
--- a/libass/dwrite_c.h
+++ b/libass/dwrite_c.h
@@ -143,6 +143,12 @@ typedef struct DWRITE_TEXT_RANGE {
 typedef struct DWRITE_TRIMMING DWRITE_TRIMMING;
 typedef struct DWRITE_UNDERLINE DWRITE_UNDERLINE;
 
+DWRITEAPI HRESULT WINAPI DWriteCreateFactory(
+  DWRITE_FACTORY_TYPE factoryType,
+  REFIID iid,
+  IUnknown **factory
+);
+
 #ifndef __MINGW_DEF_ARG_VAL
 #ifdef __cplusplus
 #define __MINGW_DEF_ARG_VAL(x) = x
